### PR TITLE
python3Packages.wandb: init at 0.12.7

### PR DESCRIPTION
**Pull request by samuela**
###### Motivation for this change
Add the Weights & Biases client API to nixpkgs.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/wandb/default.nix [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    promise
+    protobuf
+    psutil
+    python-dateutil
+    pyyaml
+    requests
+    sentry-sdk
+    shortuuid
+    yaspin
+  ] ++ lib.optionals python.isPy2 [ subprocess32 ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/wandb/default.nix:58]: py2 is no longer supported.

```suggestion
  ];
```

File: pkgs/development/python-modules/wandb/default.nix [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  };
+
+  # The wandb requirements.txt does not distinguish python2/3 dependencies. We
+  # need to drop the subprocess32 dependency when building for python3.
+  patchPhase = lib.optional python.isPy3 ''
+    sed -i 's/^subprocess32>=.*//' requirements.txt
+  '';
+
+  # git is not a setup.py dependency of wandb, but wandb does expect git to be
+  # in PATH.
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/wandb/default.nix:41]:
```suggestion
  # git is not a setup.py dependency of wandb, but wandb does expect git to be in PATH.
```

Can we patch the invocation of git instead?

File: pkgs/development/python-modules/wandb/default.nix [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    repo = "client";
+    rev = "v${version}";
+    sha256 = "08m6mcskmv4cp4zm53cqgbnbwlk7jlqsrq1gxbwzsqwngw2hcwjq";
+  };
+
+  # The wandb requirements.txt does not distinguish python2/3 dependencies. We
+  # need to drop the subprocess32 dependency when building for python3.
+  patchPhase = lib.optional python.isPy3 ''
+    sed -i 's/^subprocess32>=.*//' requirements.txt
+  '';
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/wandb/default.nix:38]:
```suggestion
  # The wandb requirements.txt does not distinguish python2/3 dependencies. We
  # need to drop the subprocess32 dependency when building for python3.
  postPatch = ''
    sed -i 's/^subprocess32>=.*//' requirements.txt
  '';
```

File: pkgs/development/python-modules/wandb/default.nix [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+, promise
+, protobuf
+, psutil
+, python
+, python-dateutil
+, pyyaml
+, requests
+, sentry-sdk
+, shortuuid
+, subprocess32 ? null
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/wandb/default.nix:19]:
```suggestion
```

File: pkgs/development/python-modules/wandb/default.nix [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  src = fetchFromGitHub {
+    owner = pname;
+    repo = "client";
+    rev = "v${version}";
+    sha256 = "08m6mcskmv4cp4zm53cqgbnbwlk7jlqsrq1gxbwzsqwngw2hcwjq";
+  };
+
+  # The wandb requirements.txt does not distinguish python2/3 dependencies. We
+  # need to drop the subprocess32 dependency when building for python3.
+  patchPhase = lib.optional python.isPy3 ''
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/wandb/default.nix:36]: Also if we would keep isPy3k is should be used as input directly.

File: pkgs/development/python-modules/wandb/default.nix (hunk 2 above)
Comment (samuela) [reply pkgs/development/python-modules/wandb/default.nix:41]: For better or worse, I believe that the git invocation is happening through `GitPython`, so if we want to patch the invocation it should happen in that derivation.

File: pkgs/development/python-modules/wandb/default.nix (hunk 1 above)
Comment (samuela) [reply pkgs/development/python-modules/wandb/default.nix:58]: Do you mean no longer supported by NixOS or no longer supported by wandb?

File: pkgs/development/python-modules/wandb/default.nix (hunk 3 above)
Comment (mkaito) [reply pkgs/development/python-modules/wandb/default.nix]: Was the `postPatch` change important?

File: pkgs/development/python-modules/yaspin/default.nix [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+}:
+
+buildPythonPackage rec {
+  pname = "yaspin";
+  version = "2.1.0";
+  format = "wheel";
+
+  # It's not straightforward to build from source since they don't include a
+  # setup.py file. The whole thing is built with poetry but adding poetry-core
+  # to nativeBuildInputs doesn't work.
```
Comment (jonringer) [inline pkgs/development/python-modules/yaspin/default.nix:14]: `format = "pyproject";` should also suffice.

Comment (samuela) [reply pkgs/development/python-modules/yaspin/default.nix]: Thanks for the pointer! Switched to building from source in https://github.com/NixOS/nixpkgs/pull/138474/commits/ace04c804e08ea402d0146b4f81d3fa94593e1e0

**Comment by samuela** [comment]
If there are no objections in the next few days, I'll go ahead and merge this.

**Comment by jonringer** [comment]
do you mind squahing the related commits? 

really should just be 2 additions

**Comment by samuela** [comment]
yes, I will squash into two commits!

**Comment by samuela** [comment]
Ok, I upgraded 0.12.2 -> 0.12.7 which allowed us to turn on the test suite now that https://github.com/wandb/client/issues/2682 has been resolved. It has also been squashed into two commits: one for yaspin and one for wandb.

If there are no objections in the next few days I'll go ahead and merge.

**Comment by YodaEmbedding** [comment]
Would love to see this merged. :)

**Comment by samuela** [comment]
@YodaEmbedding thanks for reminding! I had totally forgot about this one

**Pull request merged**